File: packages/gbnf-experiment/data/20260909T175143Z_62b61540/ported_implementation/src/rules-builder/symbol-ids.ts
export class SymbolIds {
  #map = new Map<string, number>();
  #pos = new Map<string, number>();
  #reverseMap = new Map<number, string>();

  entries(): IterableIterator<[string, number]> {
    return this.#map.entries();
  }

  [Symbol.iterator](): IterableIterator<[string, number]> {
    return this.#map.entries();
  }

  get size(): number {
    return this.#map.size;
  }

  get(key: string): number | undefined {
    return this.#map.get(key);
  }

  set(key: string, value: number, pos: number): void {
    this.#map.set(key, value);
    this.#reverseMap.set(value, key);
    this.#pos.set(key, pos);
  }

  has(key: string): boolean {
    return this.#map.has(key);
  }

  reverseGet(key: number): string {
    const value = this.#reverseMap.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return value;
  }

  getPos(key: string): number {
    const value = this.#pos.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return value;
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T125752Z_09b2c811/20260910T072624Z_d6bd66eb/ported_implementation/gbnf/rules_builder/symbol_ids.py
"""Rule name to rule id bookkeeping."""

from typing import Dict, Iterator, Optional, Tuple


class SymbolIds:
    def __init__(self) -> None:
        self.map: Dict[str, int] = {}
        self.positions: Dict[str, int] = {}
        self.reverse_map: Dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self.map)

    def __len__(self) -> int:
        return len(self.map)

    def entries(self) -> Iterator[Tuple[str, int]]:
        return iter(self.map.items())

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return iter(self.map.items())

    def get(self, key: str) -> Optional[int]:
        return self.map.get(key)

    def set(self, key: str, value: int, pos: int) -> None:
        self.map[key] = value
        self.reverse_map[value] = key
        self.positions[key] = pos

    def has(self, key: str) -> bool:
        return key in self.map

    def __contains__(self, key: str) -> bool:
        return key in self.map

    def reverse_get(self, key: int) -> str:
        val = self.reverse_map.get(key)
        if val is None:
            raise ValueError(f"SymbolIds does not contain value: {key}")
        return val

    def get_pos(self, key: str) -> int:
        val = self.positions.get(key)
        if val is None:
            raise ValueError(f"SymbolIds does not contain key: {key}")
        return val
